### PR TITLE
Implementation of Channel methods for core/db

### PR DESCRIPTION
**Pull request by toastxc**
## Please make sure to check the following tasks before opening and submitting a PR

* [x] I understand and have followed the [contribution guide](https://github.com/revoltchat/revolt/discussions/282)
* [ ] I have tested my changes locally and they are working as intended
* [ ] These changes do not have any notable side effects on other Revolt projects

### CODE REVIEW

File: crates/core/database/src/models/channels/model.rs [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+
+            /// Whether this channel is marked as not safe for work
+            #[serde(skip_serializing_if = "if_false", default)]
+            nsfw: bool,
+        },
+    }
+);
+
+auto_derived_partial!(
+    pub struct NullName {
```
Comment (insertish) [inline crates/core/database/src/models/channels/model.rs]: Is it possible to drop the public modifier here?

File: crates/core/database/src/models/channels/model.rs [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+/// Optional fields on channel object
+#[derive(Serialize, Deserialize, JsonSchema, Debug, PartialEq, Eq, Clone)]
+pub enum FieldsChannel {
+    Description,
+    Icon,
+    DefaultPermissions,
+}
+
+impl Channel {
```
Comment (insertish) [inline crates/core/database/src/models/channels/model.rs]: I would re-order the methods here to be in the order of C-R-U-D.
Create, Read, Update, Delete

I've been doing this to make it easier to consistently navigate through the models

File: crates/core/database/src/models/channels/ops.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+use crate::{revolt_result::Result, Channel, FieldsChannel, PartialChannel};
+use revolt_permissions::OverrideField;
+mod mongodb;
+mod reference;
+
+#[async_trait]
+pub trait AbstractChannels: Sync + Send {
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops.rs]: Also re-order here and in implementations

File: crates/core/permissions/Cargo.toml [hunk 4]
```diff
@@ -28,6 +28,10 @@ async-trait = "0.1.51"
 
 # Serialisation
 serde = { version = "1", features = ["derive"], optional = true }
+revolt_rocket_okapi = { version = "0.9.1", features = [ "swagger" ] }
+rocket = { version = "0.5.0-rc.2", default-features = false, features = ["json"] }
+
+bson = { version = "2.1.0" }
```
Comment (insertish) [inline crates/core/permissions/Cargo.toml]: This should be feature flagged

File: crates/core/permissions/Cargo.toml [hunk 5]
```diff
@@ -28,6 +28,10 @@ async-trait = "0.1.51"
 
 # Serialisation
 serde = { version = "1", features = ["derive"], optional = true }
+revolt_rocket_okapi = { version = "0.9.1", features = [ "swagger" ] }
```
Comment (insertish) [inline crates/core/permissions/Cargo.toml]: This is not necessary as schemars is already available behind a feature flag

File: crates/core/permissions/Cargo.toml [hunk 6]
```diff
@@ -28,6 +28,10 @@ async-trait = "0.1.51"
...
+rocket = { version = "0.5.0-rc.2", default-features = false, features = ["json"] }
```
Comment (insertish) [inline crates/core/permissions/Cargo.toml]: Is this used anywhere?

File: crates/core/permissions/src/models/server.rs [hunk 7]
```diff
@@ -1,5 +1,8 @@
+use bson::Bson;
+use revolt_rocket_okapi::{revolt_okapi::schemars, JsonSchema};
```
Comment (insertish) [inline crates/core/permissions/src/models/server.rs]: Use
```rust
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
```

However, it is necessary to add these 3 lines to lib.rs: https://github.com/revoltchat/backend/blob/master/crates/core/models/src/lib.rs#L5

File: crates/core/permissions/src/models/server.rs [hunk 8]
```diff
@@ -49,8 +81,8 @@ impl From<OverrideField> for Override {
     }
 }
 
-/*impl From<OverrideField> for Bson {
+impl From<OverrideField> for Bson {
```
Comment (insertish) [inline crates/core/permissions/src/models/server.rs]: Feature flag here after bson is changed in dependencies

File: crates/quark/src/util/variables/delta.rs [hunk 9]
```diff
@@ -41,7 +41,7 @@ pub static SMTP_FROM: Lazy<String> = Lazy::new(|| env::var("REVOLT_SMTP_FROM").u
 
 // Application Logic Settings
 pub static MAX_GROUP_SIZE: Lazy<usize> = Lazy::new(|| env::var("REVOLT_MAX_GROUP_SIZE").unwrap_or_else(|_| "50".to_string()).parse().unwrap());
-pub static MAX_BOT_COUNT: Lazy<usize> = Lazy::new(|| env::var("REVOLT_MAX_BOT_COUNT").unwrap_or_else(|_| "10".to_string()).parse().unwrap());
+pub static MAX_BOT_COUNT: Lazy<usize> = Lazy::new(|| env::var("REVOLT_MAX_BOT_COUNT").unwrap_or_else(|_| "5".to_string()).parse().unwrap());
```
Comment (insertish) [inline crates/quark/src/util/variables/delta.rs]: This seems to revert another PR.

File: crates/core/database/src/models/channels/ops/mongodb.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+            "_id": channel
+        };
+        let group_doc = doc! {
+            "$push": {
+                "recipients": user
+            }
+        };
+        Ok(self
+            .col::<Document>(COL)
+            .update_one(user_doc, group_doc, None)
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/mongodb.rs]: You can put the `doc! {}` macro straight into the method parameters, this is how I've been doing it so far on the other collections.

**Review by insertish — CHANGES_REQUESTED** [review]
Presuming the actual queries work fine, we'll be writing tests anyways so it's not a huge concern as long as it looks ok.

**Comment by toastxc** [comment]
I believe all of the requests changes have been implemented ^w^

**Comment by toastxc** [comment]
sorry I don't think I used the change request feature properly, but all the changes should be ok

File: crates/core/database/src/models/channels/ops/mongodb.rs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                    }
+                },
+                None,
+            )
+            .map(|_| ())
+            .await)
+    }
+
+    /// Insert channel role permissions
+
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/mongodb.rs]: extraneous newline

File: crates/core/database/src/models/channels/ops/reference.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
+#![allow(warnings)]
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/reference.rs]: probably shouldn't be here

File: crates/core/database/src/models/channels/ops/reference.rs [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+use super::AbstractChannels;
+use crate::ReferenceDb;
+use crate::{Channel, FieldsChannel, MongoDb, PartialChannel};
+use futures::{FutureExt, StreamExt};
+use revolt_permissions::OverrideField;
+use revolt_result::Result;
+static COL: &str = "channels";
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/reference.rs]: This line can be omitted here.

File: crates/core/database/src/models/channels/ops/reference.rs [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        let mut channels = self.channels.lock().await;
+        channels
+            .get(user_id)
+            .cloned()
+            .ok_or_else(|| create_database_error!("fetch", "channel"))
+    }
+
+    // Fetch direct message channel (PMs)
+    async fn find_direct_message_channel(&self, user_a: &str, user_b: &str) -> Result<Channel> {
+        todo!()
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/reference.rs]: impl

File: crates/core/database/src/models/channels/ops/reference.rs [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    // Update channel
+    async fn update_channel(
+        &self,
+        id: &str,
+        channel: &PartialChannel,
+        remove: Vec<FieldsChannel>,
+    ) -> Result<()> {
+        todo!()
```
Comment (insertish) [inline crates/core/database/src/models/channels/ops/reference.rs]: impl

File: crates/delta/src/routes/servers/permissions_set_default.rs [hunk 16]
```diff
@@ -1,28 +1,17 @@
...
-#[derive(Deserialize, JsonSchema)]
-pub struct DataSetServerDefaultPermission {
-    /// Default member permission value
-    permissions: u64,
-}
-
-/// # Set Default Permission
-///
-/// Sets permissions for the default role in this server.
 #[openapi(tag = "Server Permissions")]
```
Comment (insertish) [inline crates/delta/src/routes/servers/permissions_set_default.rs]: Route description removed by accident

**Review by insertish — CHANGES_REQUESTED** [review]
Looks almost ready for merge, found a few more things

**Comment by toastxc** [comment]
There was a conflict, earlier I added Channel as a dependancy for the DB MutexGuard.

**Comment by insertish** [comment]
Webhooks just got merged, now conflicting, should be fairly quick to fix.

**Pull request closed without merging**